forwardingproxy.go

```diff
@@ -0,0 +1,224 @@
+package tunnel
+
+import (
+	"bufio"
+	"encoding/base64"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/http/httputil"
+	"strings"
+	"time"
+
+	"github.com/mmatczuk/go-http-tunnel/log"
+	"github.com/mmatczuk/go-http-tunnel/proto"
+)
+
+// ForwardingProxy uses http tunnel.
+type ForwardingProxy struct {
+	// localAddr specifies default TCP address of the local server.
+	localAddr string
+	// localAddrMap specifies mapping from ControlMessage.ForwardedHost to
+	// local server address, keys may contain host and port, only host or
+	// only port. The order of precedence is the following
+	// * host and port
+	// * port
+	// * host
+	localAddrMap map[string]string
+	// logger is the proxy logger.
+	logger              log.Logger
+	AuthUser            string
+	AuthPass            string
+	ForwardingHTTPProxy *httputil.ReverseProxy
+	DestDialTimeout     time.Duration
+	DestReadTimeout     time.Duration
+	DestWriteTimeout    time.Duration
+	ClientReadTimeout   time.Duration
+	ClientWriteTimeout  time.Duration
+}
+
+// NewForwardingProxy creates new direct TCPProxy, everything will be proxied to
+// localAddr.
+func NewForwardingProxy(localAddr string, logger log.Logger) *ForwardingProxy {
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+
+	return &ForwardingProxy{
+		localAddr: localAddr,
+		logger:    logger,
+	}
+}
+
+// NewMultiForwardingProxy creates a new dispatching TCPProxy, connections may go to
+// different backends based on localAddrMap.
+func NewMultiForwardingProxy(localAddrMap map[string]string, logger log.Logger) *ForwardingProxy {
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+
+	return &ForwardingProxy{
+		localAddrMap:        localAddrMap,
+		logger:              logger,
+		ForwardingHTTPProxy: NewReverseProxy(),
+		AuthUser:            "",
+		AuthPass:            "",
+		DestDialTimeout:     10 * time.Second,
+		DestReadTimeout:     5 * time.Second,
+		DestWriteTimeout:    5 * time.Second,
+		ClientReadTimeout:   5 * time.Second,
+		ClientWriteTimeout:  5 * time.Second,
+	}
+}
+
+// Proxy is a ProxyFunc.
+func (p *ForwardingProxy) Proxy(w io.Writer, r io.ReadCloser, msg *proto.ControlMessage) {
+	switch msg.ForwardedProto {
+	case proto.HTTPCONNECT:
+		// ok
+	default:
+		p.logger.Log(
+			"level", 0,
+			"msg", "unsupported protocol",
+			"ctrlMsg", msg,
+		)
+		return
+	}
+
+	rw, ok := w.(http.ResponseWriter)
+	if !ok {
+		p.logger.Log(
+			"level", 0,
+			"msg", "expected http.ResponseWriter",
+			"ctrlMsg", msg,
+		)
+	}
+
+	req, err := http.ReadRequest(bufio.NewReader(r))
+	if err != nil {
+		p.logger.Log(
+			"level", 0,
+			"msg", "failed to read request",
+			"ctrlMsg", msg,
+			"err", err,
+		)
+		return
```

Comment:
likewise, do we need a response to the caller to let them know the call failed?

Reply:
Possibly, but again is going to timeout.

Reply:
okay